### PR TITLE
Add New Integration for DeepInfra Embedding Model

### PR DESCRIPTION
**Pull request by ovuruska**
**Description**
This pull request introduces a new integration for the DeepInfra Inference API.

**Motivation and Context**
Users are able to use DeepInfra's models using llama_index.

**Fixes**
- No specific issue fixed; this is a new feature addition.

**New Package?**
- [x] Yes
- [ ] No

If yes, I have filled in the `tool.llamahub` section in the `pyproject.toml` and provided a detailed `README.md` for my new integration.

**Version Bump?**
- [x] Yes
- [ ] No

**Type of Change**
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**
- Tests have been added to verify the functionality of the new embedding model integration. These tests ensure that both synchronous and asynchronous methods perform as expected.
- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

**Suggested Checklist:**
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

**Comment by ichernev** [comment]
I don't see the code that does the auto-batching (the one you added for lang-chain). Other than that it looks like a standalone project, which is probably how this is organized.

Other than that looks good.

File: llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/llama_index/embeddings/deepinfra/base.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    def _get_query_embedding(self, query: str) -> List[float]:
+        """
+        Get query embedding.
+        """
+        return self._post([query])["embeddings"][0]
+
+    async def _aget_query_embedding(self, query: str) -> List[float]:
+        """
+        Async get query embedding.
```
Comment (ichernev) [inline llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/llama_index/embeddings/deepinfra/base.py:150]: Async here? We do some sharing of sync and async code by computing headers, url and body in methods and calling those in sync and async requests/HTTP client APIs

**Comment by ichernev** [comment]
Also implement sync and async.

Also what is the difference between text and query? Actually some models do make a difference (normally you prefix the text with `query:` or `prompt:` or sth similar)

File: llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/BUILD [hunk 2]
```diff
@@ -0,0 +1,3 @@
+poetry_requirements(
+    name="poetry",
```
Comment (ichernev) [inline llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/BUILD:2]: Are you sure you need this? I mean if you already have poetry_requirements that would require poetry to read, no?

File: llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/README.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+In order to provide the API token, you can set the `DEEPINFRA_API_TOKEN` environment variable or pass it as an argument to the `DeepinfraEmbeddings` class.
+
+### Use with default configuration
+
+```python
+from dotenv import load_dotenv, find_dotenv
+from llama_index.embeddings.deepinfra import DeepInfraEmbeddingModel
+
+_ = load_dotenv(find_dotenv())
```
Comment (ichernev) [inline llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/README.md]: Maybe a comment to explain what are you loading exactly, not obvious to me.

File: llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/README.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    model_id="BAAI/bge-large-en-v1.5",
+    query_prefix="query: ",
+)
+
+response = model.get_query_embedding("hello world")
+# Print the embeddings
+print(response)
+```
+
+### Use with text prefix
```
Comment (ichernev) [inline llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/README.md]: There should be one example which passes both query and text prefixes, with a short description on why that would be useful. I don't see how having two examples is better.

File: llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/llama_index/embeddings/deepinfra/base.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        """
+        Async get text embeddings.
+        """
+        return await self._apost(self._add_text_prefix(texts))
+
+    def _add_query_prefix(self, queries: List[str]) -> List[str]:
+        """
+        Add query prefix to queries.
+        """
+        return [self._query_prefix + query for query in queries]
```
Comment (ichernev) [inline llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/llama_index/embeddings/deepinfra/base.py]: You can do `[self._query_prefix + query for query in queries] if self._query_prefix else queries` so it doesn't slow things down without query prefix

File: llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/README.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+)
+
+texts = ["hello world", "goodbye world"]
+
+response = model.get_text_embedding_batch(texts)
+# Print the embeddings
+print(response)
+```
+
+### Asynchronous requests
```
Comment (ichernev) [inline llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/README.md]: I guess if all others have these verbose examples then fine, but the setup is the same, only the method call is different, and even that is the API of the BaseEmbedding (i.e nothing deepinfra specific). So you decide if you keep this or shorten it a bit (maybe a single setup + multiple fn calls).

File: llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/llama_index/embeddings/deepinfra/base.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        Add text prefix to texts.
+        """
+        return [self._text_prefix + text for text in texts]
+
+
+def _chunk(items: List[str]) -> List[List[str]]:
+    """
+    Chunk items into batches of size MAX_BATCH_SIZE.
+    """
+    return [items[i : i + MAX_BATCH_SIZE] for i in range(0, len(items), MAX_BATCH_SIZE)]
```
Comment (ichernev) [inline llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/llama_index/embeddings/deepinfra/base.py]: `MAX_BATCH_SIZE` can be a constructor arg, and this function can be a method

Comment (ovuruska) [reply llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/llama_index/embeddings/deepinfra/base.py]: Isn't this parameter related with deepinfra's backend?

File: llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/README.md (hunk 6 above)
Comment (ovuruska) [reply llama-index-integrations/embeddings/llama-index-embeddings-deepinfra/README.md]: Dealt with it. Decreased the code repetition.

**Comment by ichernev** [comment]
Looks good, send it upstream!

**Pull request merged**